src/Hl7.Fhir.Specification/Specification/Snapshot/SnapshotGenerator.cs

```diff
@@ -757,6 +763,29 @@ private async T.Task createNewElement(ElementDefinitionNavigator snap, ElementDe
             await mergeElement(snap, diff).ConfigureAwait(false);
         }
 
+        private void removeNewTypeConstraint(ElementDefinition element, StructureDefinition typeStructure)
+        {
+            if (typeStructure?.Differential?.Element != null && !element.Constraint.IsNullOrEmpty())
+            {
+                List<ElementDefinition.ConstraintComponent> newConstraints = null;
+               
+                //See if there are new constraints introduced by the type
+                var nav = new ElementDefinitionNavigator(typeStructure.Differential.Element);
+                if (nav.MoveToFirstChild())
+                {
+                    if(nav.Current.IsRootElement()) 
+                        newConstraints = nav.Current.Constraint;
+                }
+                //If there are any newly introduced constraints, remove them from the new element.
+                if (newConstraints != null)
+                {
+                    var keys = newConstraints.Select(c => c.Key);
+                    var removedConstraints = element.Constraint.Where(c => keys.Contains(c.Key))?.ToList();
```

Comment:
Another way of doing this, is:
```c#
element.Constraint.RemoveAll(c => keys.Contains(c.Key));
```

Although with big lists you should be carefull, but that is here not the case: more than 1000 constraints would be not realistic. When you deal with large list you could also do this:
```c#
var keys  = new HashSet<string>(newConstraints.Select(c => c.Key));
element.Constraint.RemoveAll(c => removedConstraints.Contains(c.Key));
```
That makes the complexity O(n+m) instead of O(n*m).

Reply:
Thanks, I'll fix that!